### PR TITLE
Propagate fatal errors from background tasks 

### LINKED ISSUE #2324: Propagate fatal errors from background tasks
**Issue summary**

Forest uses a bunch of background tasks but these tasks rarely propagate fatal errors to the main task. This causes fatal to put Forest into a non-working, zombie state where it keeps running but doesn't do any processing.

Steps to reproduce:
1. Download https://forest-snapshots.fra1.digitaloceanspaces.com/forest_snapshot_calibnet_2022-12-05_height_95741.car
2. Import the snapshot: `forest --chain calibnet --encrypt-keystore false --import-snapshot forest_snapshot_calibnet_2022-12-05_height_95741.car`
3. Notice that it complains about `Invalid tipset: no msg root with cid bafy2bzaceb7yrvmubpzoltkhytajykrgbhfpgmj4yqg4nlnwqzcsnfja4zvj6` and then stops validating tipsets.
4. The failure happens here: https://github.com/ChainSafe/forest/blob/1132fe205e9dc6f27b833d855f51631e5b8958c2/blockchain/message_pool/src/msgpool/msg_pool.rs#L242

@creativcoder has suggested a few ways of propagating fatal errors from background tasks to the main task:
- https://doc.rust-lang.org/stable/std/sync/struct.Condvar.html
- Passing a Sender<ToParent> to the worker thread.
- https://docs.rs/tokio/latest/tokio/sync/struct.Notify.html#method.notify_waiters

Tasks:
- [ ] Evaluate the different possible solutions and pick the one you feel is best.
- [ ] Propagate errors from the message pool to the main thread such that loading a broken snapshot fails with a proper error message.
- [ ] Identifier other background tasks that may fail without propagating their errors and open a new issue to track them.

**Other information and links**

### PR DESCRIPTION
**Pull request by tyshko5**
**Summary of changes**
Changes introduced in this pull request:
- 

**Reference issue to close (if applicable)**

Closes #2324

**Other information and links**

### CODE REVIEW

File: forest/daemon/src/daemon.rs [hunk 1]
```diff
@@ -377,6 +380,15 @@ pub(super) async fn start(config: Config, detached: bool) -> Db {
...
+        if let Ok(res_inner) = res {
+            if let Err(error_message) = res_inner {
+                let msg = format!("services failure: {}", error_message);
+                cli_error_and_die(msg, 1);
+            }
+        }
+    }
+
     // Block until ctrl-c is hit
     ctrlc_oneshot.await.unwrap();
```
Comment (lemmih) [inline forest/daemon/src/daemon.rs:393]: I don't think this works. You're no longer listening for ctrl-c.

**Review by lemmih — CHANGES_REQUESTED** [review]

**Comment by lemmih** [comment]
You still have to handle ctrl-c correctly.

Also, show the new output for running forest on the broken snapshot.

File: forest/daemon/src/daemon.rs [hunk 2]
```diff
@@ -169,6 +169,7 @@ pub(super) async fn start(config: Config, detached: bool) -> Db {
             {
                 error!("Failed to initiate prometheus server: {e}");
             }
+            Ok(())
```
Comment (lemmih) [inline forest/daemon/src/daemon.rs:172]: Why ignore the error here?

File: forest/daemon/src/daemon.rs [hunk 3]
```diff
@@ -337,6 +339,7 @@ pub(super) async fn start(config: Config, detached: bool) -> Db {
                 FOREST_VERSION_STRING.as_str(),
             )
             .await;
+            Ok(())
```
Comment (lemmih) [inline forest/daemon/src/daemon.rs:342]: Why ignore the error message here?

File: forest/daemon/src/daemon.rs [hunk 4]
```diff
@@ -306,6 +307,7 @@ pub(super) async fn start(config: Config, detached: bool) -> Db {
     let sync_state = chain_muxer.sync_state_cloned();
     services.spawn(async {
         chain_muxer.await;
+        Ok(())
```
Comment (lemmih) [inline forest/daemon/src/daemon.rs:310]: Why ignore the error message here?

**Comment by tyshko5** [comment]
> You still have to handle ctrl-c correctly.
> 
> Also, show the new output for running forest on the broken snapshot.

Trying to find a way to fix ctrl-c

File: forest/daemon/src/daemon.rs (hunk 1 above)
Comment (tyshko5) [reply forest/daemon/src/daemon.rs]: I think I fixed it

File: forest/daemon/src/daemon.rs (hunk 2 above)
Comment (tyshko5) [reply forest/daemon/src/daemon.rs:176]: It's bad prometheus server failed to initiate, but it seems to to be critical to Forest

File: forest/daemon/src/daemon.rs (hunk 3 above)
Comment (tyshko5) [reply forest/daemon/src/daemon.rs]: Fixed

File: forest/daemon/src/daemon.rs (hunk 4 above)
Comment (tyshko5) [reply forest/daemon/src/daemon.rs]: Fixed

**Comment by lemmih** [comment]
I don't think that's a great way to deal with ctrl-c. You would be better off using `select!`.

**Comment by tyshko5** [comment]
> I don't think that's a great way to deal with ctrl-c. You would be better off using `select!`.

Why? Currently the interface is unified recreating the old logic completely. With select as far as I understand the situation is back to what it was.

File: forest/daemon/src/daemon.rs [hunk 5]
```diff
@@ -361,11 +363,6 @@ pub(super) async fn start(config: Config, detached: bool) -> Db {
 
     select! {
         () = sync_from_snapshot(&config, &state_manager).fuse() => {},
-        _ = ctrlc_oneshot => {
-            // Cancel all async services
-            services.shutdown().await;
-            return db;
-        },
```
Comment (lemmih) [inline forest/daemon/src/daemon.rs:368]: You can't just delete this code. It was here for a reason.

Comment (lemmih) [reply forest/daemon/src/daemon.rs:368]: The code is required for a graceful shutdown. We _do_ want a graceful shutdown. That's the entire point of the ctrl-c handler.

**Comment by tyshko5** [comment]
> > > I don't think that's a great way to deal with ctrl-c. You would be better off using `select!`.
> > 
> > 
> > Why? Currently the interface is unified recreating the old logic completely. With select as far as I understand the situation is back to what it was.
> 
> Imagine if one of the background services terminates early. You would interpret that as a ctrl-c and quit Forest. It's not intuitive for Forest developers that ending a background thread early is the same as a ctrl-c. And it's also not intuitive behavior for an end-user.

In case of ctrl-c, I can still return a specific error message, removing the current else clause with `break`. 
Shouldn't that work, or I'm missing something?

**Comment by lemmih** [comment]
> > > > I don't think that's a great way to deal with ctrl-c. You would be better off using `select!`.
> > > 
> > > 
> > > Why? Currently the interface is unified recreating the old logic completely. With select as far as I understand the situation is back to what it was.
> > 
> > 
> > Imagine if one of the background services terminates early. You would interpret that as a ctrl-c and quit Forest. It's not intuitive for Forest developers that ending a background thread early is the same as a ctrl-c. And it's also not intuitive behavior for an end-user.
> 
> In case of ctrl-c, I can still return a specific error message, removing the current else clause with `break`. Shouldn't that work, or I'm missing something?

Then, conceptually, the background services would have three possible return values: `Ok`, `Err`, and `CtrlCInterrupt`. I don't doubt you could make it work but I'm unconvinced that it is a good solution.

Why not just use `select!`? Then you wouldn't need to introduce any new return values and the control-flow would be much easier to see in the code.

File: forest/daemon/src/daemon.rs (hunk 5 above)
Comment (tyshko5) [reply forest/daemon/src/daemon.rs:368]: We are still doing a graceful shutdown using the same commands in the end of the `start` function

Comment (lemmih) [reply forest/daemon/src/daemon.rs:368]: If ctrl-c is hit duing `sync_from_snapshot`, how do we do a graceful shutdown?

Comment (lemmih) [reply forest/daemon/src/daemon.rs:368]: Keep in mind that `sync_from_snapshot` may take 10 minutes to complete.

Comment (tyshko5) [reply forest/daemon/src/daemon.rs]: Reverted

File: forest/daemon/src/daemon.rs (hunk 1 above)
Comment (tyshko5) [reply forest/daemon/src/daemon.rs]: Refixed using `select!`

File: forest/daemon/src/daemon.rs [hunk 6]
```diff
@@ -377,8 +377,22 @@ pub(super) async fn start(config: Config, detached: bool) -> Db {
...
     services.spawn(p2p_service.run());
 
-    // Block until ctrl-c is hit
-    ctrlc_oneshot.await.unwrap();
+    loop {
+        select! {
+            option = services.join_next().fuse() => {
+                if let Some(res) = option {
+                    if let Ok(res_inner) = res {
+                        if let Err(error_message) = res_inner {
```
Comment (creativcoder) [inline forest/daemon/src/daemon.rs]: if we're not handling other clauses for the above two matches, I think we we can do: 
```
if let Some(Ok(Err(error_message)) = option {

}
```

File: forest/daemon/src/daemon.rs (hunk 2 above)
Comment (lemmih) [reply forest/daemon/src/daemon.rs:172]: We already fail if bind to the port or find the DB path. Let's fail as eagerly as possible and exit Forest if something goes wrong in prometheus. It'll then be prometheus' responsibility to properly deal with non-fatal errors.

File: forest/daemon/src/daemon.rs [hunk 7]
```diff
@@ -377,8 +377,20 @@ pub(super) async fn start(config: Config, detached: bool) -> Db {
...
+                    let msg = format!("services failure: {}", error_message);
+                    error!("{}", msg);
+                    break
+                }
+            },
+            complete => select! {
+                _ = ctrlc_oneshot => break
+            },
+        }
+    }
```
Comment (lemmih) [inline forest/daemon/src/daemon.rs:393]: I'm 90% sure that this code does not work.

File: blockchain/chain_sync/src/consensus.rs [hunk 8]
```diff
@@ -111,7 +111,7 @@ pub trait Proposer {
         state_manager: Arc<StateManager<DB>>,
         mpool: Arc<MP>,
         submitter: SyncGossipSubmitter,
-        services: &mut JoinSet<()>,
+        services: &mut JoinSet<Result<(), String>>,
```
Comment (hanabi1224) [inline blockchain/chain_sync/src/consensus.rs:114]: how about `anyhow::Result<()>`?

File: forest/daemon/src/daemon.rs (hunk 2 above)
Comment (tyshko5) [reply forest/daemon/src/daemon.rs]: Fixed accordingly

File: forest/daemon/src/daemon.rs (hunk 7 above)
Comment (tyshko5) [reply forest/daemon/src/daemon.rs]: Changed this logic

File: blockchain/chain_sync/src/consensus.rs (hunk 8 above)
Comment (tyshko5) [reply blockchain/chain_sync/src/consensus.rs]: Fixed accordingly

File: forest/daemon/src/daemon.rs [hunk 9]
```diff
@@ -167,7 +167,12 @@ pub(super) async fn start(config: Config, detached: bool) -> Db {
...
                 forest_metrics::init_prometheus(prometheus_listener, db_directory, db).await
             {
-                error!("Failed to initiate prometheus server: {e}");
+                Err(anyhow::anyhow!(
+                    "{}",
+                    format!("Failed to initiate prometheus server: {e}")
+                ))
+            } else {
+                Ok(())
             }
```
Comment (lemmih) [inline forest/daemon/src/daemon.rs:176]: This can be refactored.

File: forest/daemon/src/daemon.rs [hunk 10]
```diff
@@ -377,8 +381,28 @@ pub(super) async fn start(config: Config, detached: bool) -> Db {
...
+            option = services.join_next().fuse() => {
+                if let Some(Ok(Err(error_message))) = option {
+                    let msg = format!("services failure: {}", error_message);
+                    error!("{}", msg);
+                    break
+                }
+            },
+            _ = ctrlc_oneshot => break
+        }
+    }
```
Comment (lemmih) [inline forest/daemon/src/daemon.rs:399]: This logic is still wrong. We want to block until an error has occurred or ctrl-c has been pressed. Let's discuss if this is still unclear.

File: node/forest_libp2p/src/service.rs [hunk 11]
```diff
@@ -228,12 +228,9 @@ where
...
         let mut interval =
             IntervalStream::new(tokio::time::interval(Duration::from_secs(15))).fuse();
-        let pubsub_block_str = format!("{}/{}", PUBSUB_BLOCK_STR, self.network_name);
-        let pubsub_msg_str = format!("{}/{}", PUBSUB_MSG_STR, self.network_name);
 
-        let mut hello_request_table = HashMap::new();
-        let mut cx_request_table = HashMap::new();
-        let mut outgoing_bitswap_query_ids = HashMap::new();
+        let (mut hello_request_table, mut cx_request_table, mut outgoing_bitswap_query_ids) =
+            (HashMap::new(), HashMap::new(), HashMap::new());
```
Comment (lemmih) [inline node/forest_libp2p/src/service.rs:233]: Try changing as few lines as possible.

File: blockchain/chain_sync/src/consensus.rs [hunk 12]
```diff
@@ -111,7 +111,7 @@ pub trait Proposer {
         state_manager: Arc<StateManager<DB>>,
         mpool: Arc<MP>,
         submitter: SyncGossipSubmitter,
-        services: &mut JoinSet<()>,
+        services: &mut JoinSet<Result<(), anyhow::Error>>,
```
Comment (lemmih) [inline blockchain/chain_sync/src/consensus.rs:114]:
```suggestion
        services: &mut JoinSet<anyhow::Result<()>>,
```

Comment (tyshko5) [reply blockchain/chain_sync/src/consensus.rs]: Fixed

File: forest/daemon/src/daemon.rs [hunk 13]
```diff
@@ -392,6 +403,24 @@ pub(super) async fn start(config: Config, detached: bool) -> Db {
...
 
+async fn propagate_error(services: &mut JoinSet<Result<(), anyhow::Error>>) -> anyhow::Error {
+    while !services.is_empty() {
+        select! {
+            option = services.join_next().fuse() => {
+                if let Some(Ok(Err(error_message))) = option {
+                    return error_message
+                } else {
+                    continue
+                }
```
Comment (lemmih) [inline forest/daemon/src/daemon.rs:414]: No need for `continue` here.

File: node/forest_libp2p/src/service.rs [hunk 14]
```diff
@@ -258,7 +258,7 @@ where
                             &pubsub_block_str,
                             &pubsub_msg_str,).await;
                     },
-                    None => { break; },
+                    None => { break Ok(()); },
```
Comment (lemmih) [inline node/forest_libp2p/src/service.rs:261]: How about moving the `Ok(())` such that you only write it once?

File: blockchain/consensus/deleg_cns/src/composition.rs [hunk 15]
```diff
@@ -27,7 +27,7 @@ pub async fn consensus<DB, MP>(
     keystore: &Arc<RwLock<KeyStore>>,
     mpool: &Arc<MP>,
     submitter: SyncGossipSubmitter,
-    services: &mut JoinSet<()>,
+    services: &mut JoinSet<Result<(), anyhow::Error>>,
```
Comment (lemmih) [inline blockchain/consensus/deleg_cns/src/composition.rs:30]: Be consistent in using `anyhow::Result<()>`

File: forest/daemon/src/daemon.rs (hunk 9 above)
Comment (tyshko5) [reply forest/daemon/src/daemon.rs]: Done

File: node/forest_libp2p/src/service.rs (hunk 14 above)
Comment (tyshko5) [reply node/forest_libp2p/src/service.rs]: Done

File: forest/daemon/src/daemon.rs (hunk 13 above)
Comment (tyshko5) [reply forest/daemon/src/daemon.rs]: Removed

File: blockchain/consensus/deleg_cns/src/proposer.rs [hunk 16]
```diff
@@ -102,15 +102,20 @@ impl Proposer for DelegatedProposer {
...
         services.spawn(async move {
             if let Err(e) = self.run(state_manager, mpool.as_ref(), &submitter).await {
-                error!("block proposal stopped: {}", e)
+                Err(anyhow::anyhow!(
+                    "{}",
+                    format!("block proposal stopped: {}", e)
+                ))
+            } else {
+                Ok(())
             }
```
Comment (lemmih) [inline blockchain/consensus/deleg_cns/src/proposer.rs:119]: See `map_err` or just use '?`.

File: blockchain/message_pool/src/msgpool/msg_pool.rs [hunk 17]
```diff
@@ -239,13 +239,20 @@ where
...
                         .await
-                        .unwrap_or_else(|err| warn!("Error changing head: {:?}", err));
+                        .map_err(|err| err.to_string());
+
+                        if let Err(err) = res {
+                            break Err(anyhow::anyhow!(
+                                "{}",
+                                format!("Error changing head: {:?}", err)
+                            ));
+                        }
```
Comment (lemmih) [inline blockchain/message_pool/src/msgpool/msg_pool.rs:249]: Can this be refactored to be more idiomatic? Can you use `?`?

File: forest/daemon/src/daemon.rs [hunk 18]
```diff
@@ -164,11 +166,11 @@ pub(super) async fn start(config: Config, detached: bool) -> Db {
...
         services.spawn(async {
-            if let Err(e) =
-                forest_metrics::init_prometheus(prometheus_listener, db_directory, db).await
-            {
-                error!("Failed to initiate prometheus server: {e}");
-            }
+            forest_metrics::init_prometheus(prometheus_listener, db_directory, db)
+                .await
+                .map_err(|e| {
+                    anyhow::anyhow!("{}", format!("Failed to initiate prometheus server: {e}"))
```
Comment (lemmih) [inline forest/daemon/src/daemon.rs:172]: Avoid formatting twice.

File: blockchain/consensus/deleg_cns/src/proposer.rs [hunk 19]
```diff
@@ -102,15 +102,20 @@ impl Proposer for DelegatedProposer {
...
     where
         DB: Blockstore + Store + Clone + Sync + Send + 'static,
         MP: MessagePoolApi + Send + Sync + 'static,
     {
         services.spawn(async move {
             if let Err(e) = self.run(state_manager, mpool.as_ref(), &submitter).await {
-                error!("block proposal stopped: {}", e)
+                Err(anyhow::anyhow!(
+                    "{}",
+                    format!("block proposal stopped: {}", e)
```
Comment (lemmih) [inline blockchain/consensus/deleg_cns/src/proposer.rs:115]: Avoid formatting twice.

File: blockchain/message_pool/src/msgpool/msg_pool.rs [hunk 20]
```diff
@@ -239,13 +239,20 @@ where
...
                     }
                     Err(RecvError::Lagged(e)) => {
                         warn!("Head change subscriber lagged: skipping {} events", e);
                     }
                     Err(RecvError::Closed) => {
-                        break;
+                        break Ok::<(), anyhow::Error>(());
```
Comment (lemmih) [inline blockchain/message_pool/src/msgpool/msg_pool.rs:255]: Is the type annotation required?

File: blockchain/message_pool/src/msgpool/selection.rs [hunk 21]
```diff
@@ -693,7 +693,9 @@ mod test_selection {
 
     const TEST_GAS_LIMIT: i64 = 6955002;
 
-    async fn make_test_mpool(joinset: &mut JoinSet<()>) -> MessagePool<TestApi> {
+    async fn make_test_mpool(
+        joinset: &mut JoinSet<Result<(), anyhow::Error>>,
```
Comment (lemmih) [inline blockchain/message_pool/src/msgpool/selection.rs:697]: Be consistent about `anyhow::Result`.

File: node/forest_libp2p/src/service.rs [hunk 22]
```diff
@@ -216,7 +216,7 @@ where
     }
 
     /// Starts the libp2p service networking stack. This Future resolves when shutdown occurs.
-    pub async fn run(mut self) {
+    pub async fn run(mut self) -> Result<(), anyhow::Error> {
         info!("Running libp2p service");
         Swarm::listen_on(&mut self.swarm, self.config.listening_multiaddr).unwrap();
```
Comment (lemmih) [inline node/forest_libp2p/src/service.rs:221]: Can we remove this `unwrap`?

File: blockchain/consensus/deleg_cns/src/proposer.rs (hunk 19 above)
Comment (tyshko5) [reply blockchain/consensus/deleg_cns/src/proposer.rs]: Fixed

File: forest/daemon/src/daemon.rs [hunk 23]
```diff
@@ -392,6 +398,22 @@ pub(super) async fn start(config: Config, detached: bool) -> Db {
...
+    while !services.is_empty() {
+        select! {
+            option = services.join_next().fuse() => {
+                if let Some(Ok(Err(error_message))) = option {
+                    return error_message
+                }
+            },
+        }
+    }
+    sleep(Duration::new(3200000000, 0));
```
[the remark this answers is not in the document]
Comment (tyshko5) [reply forest/daemon/src/daemon.rs]: Changed to `tokio` analogue

File: forest/daemon/src/daemon.rs [hunk 24]
```diff
@@ -377,8 +378,13 @@ pub(super) async fn start(config: Config, detached: bool) -> Db {
 
     services.spawn(p2p_service.run());
 
-    // Block until ctrl-c is hit
-    ctrlc_oneshot.await.unwrap();
+    select! {
+        err = propagate_error(&mut services).fuse() => {
+            let msg = format!("services failure: {}", err);
```
[the remark this answers is not in the document]
Comment (tyshko5) [reply forest/daemon/src/daemon.rs]: Fixed

File: forest/daemon/src/daemon.rs (hunk 18 above)
Comment (tyshko5) [reply forest/daemon/src/daemon.rs]: Fixed

File: blockchain/consensus/deleg_cns/src/proposer.rs (hunk 16 above)
Comment (tyshko5) [reply blockchain/consensus/deleg_cns/src/proposer.rs]: Fixed

File: blockchain/message_pool/src/msgpool/selection.rs (hunk 21 above)
Comment (tyshko5) [reply blockchain/message_pool/src/msgpool/selection.rs]: Fixed

File: node/forest_libp2p/src/service.rs (hunk 22 above)
Comment (tyshko5) [reply node/forest_libp2p/src/service.rs]: Removed

File: blockchain/message_pool/src/msgpool/msg_pool.rs [hunk 25]
```diff
@@ -239,13 +239,13 @@ where
                             app,
                         )
                         .await
-                        .unwrap_or_else(|err| warn!("Error changing head: {:?}", err));
+                        .map_err(|err| anyhow::anyhow!("Error changing head: {:?}", err))?;
```
Comment (lemmih) [inline blockchain/message_pool/src/msgpool/msg_pool.rs:242]: [context](https://docs.rs/anyhow/latest/anyhow/trait.Context.html#tymethod.context) looks appropriate here.

File: forest/daemon/src/daemon.rs [hunk 26]
```diff
@@ -143,7 +144,7 @@ pub(super) async fn start(config: Config, detached: bool) -> Db {
 
     let db = open_db(&config);
 
-    let mut services = JoinSet::new();
+    let mut services: JoinSet<Result<(), anyhow::Error>> = JoinSet::new();
```
Comment (lemmih) [inline forest/daemon/src/daemon.rs:147]: `anyhow::Result`?

File: forest/daemon/src/daemon.rs [hunk 27]
```diff
@@ -164,11 +165,9 @@ pub(super) async fn start(config: Config, detached: bool) -> Db {
...
         let db = db.clone();
         services.spawn(async {
-            if let Err(e) =
-                forest_metrics::init_prometheus(prometheus_listener, db_directory, db).await
-            {
-                error!("Failed to initiate prometheus server: {e}");
-            }
+            forest_metrics::init_prometheus(prometheus_listener, db_directory, db)
+                .await
+                .map_err(|e| anyhow::anyhow!("Failed to initiate prometheus server: {e}"))
```
Comment (lemmih) [inline forest/daemon/src/daemon.rs:170]: See [context](https://docs.rs/anyhow/latest/anyhow/trait.Context.html#tymethod.context)

File: forest/daemon/src/daemon.rs [hunk 28]
```diff
@@ -392,6 +394,22 @@ pub(super) async fn start(config: Config, detached: bool) -> Db {
...
+    while !services.is_empty() {
+        select! {
+            option = services.join_next().fuse() => {
+                if let Some(Ok(Err(error_message))) = option {
+                    return error_message
+                }
+            },
+        }
+    }
+    tokio::time::sleep(Duration::new(3200000000, 0)).await;
```
Comment (lemmih) [inline forest/daemon/src/daemon.rs:407]: Does `tokio::time::sleep` have a max sleeping duration? What happens if you exceed the max duration?

File: forest/daemon/src/daemon.rs [hunk 29]
```diff
@@ -392,6 +394,22 @@ pub(super) async fn start(config: Config, detached: bool) -> Db {
...
+    anyhow::Error::msg(
```
Comment (lemmih) [inline forest/daemon/src/daemon.rs:408]: `bail!`?

File: node/forest_libp2p/src/service.rs [hunk 30]
```diff
@@ -216,9 +216,10 @@ where
     }
 
     /// Starts the libp2p service networking stack. This Future resolves when shutdown occurs.
-    pub async fn run(mut self) {
+    pub async fn run(mut self) -> Result<(), anyhow::Error> {
```
Comment (lemmih) [inline node/forest_libp2p/src/service.rs:219]: `anyhow::Result`?

File: node/forest_libp2p/src/service.rs [hunk 31]
```diff
@@ -216,9 +216,10 @@ where
...
         info!("Running libp2p service");
-        Swarm::listen_on(&mut self.swarm, self.config.listening_multiaddr).unwrap();
+        Swarm::listen_on(&mut self.swarm, self.config.listening_multiaddr)
+            .map_err(|e| anyhow::anyhow!("{e}"))?;
```
Comment (lemmih) [inline node/forest_libp2p/src/service.rs:222]: Is this needed?

File: forest/daemon/src/daemon.rs [hunk 32]
```diff
@@ -377,8 +375,12 @@ pub(super) async fn start(config: Config, detached: bool) -> Db {
 
     services.spawn(p2p_service.run());
 
-    // Block until ctrl-c is hit
-    ctrlc_oneshot.await.unwrap();
+    select! {
+        err = propagate_error(&mut services).fuse() => {
+            error!("services failure: {}", err);
+        },
```
Comment (lemmih) [inline forest/daemon/src/daemon.rs:381]:
```suggestion
        err = propagate_error(&mut services).fuse() => error!("services failure: {err}"),
```

**Review by lemmih — CHANGES_REQUESTED** [review]

File: node/forest_libp2p/src/service.rs (hunk 31 above)
Comment (tyshko5) [reply node/forest_libp2p/src/service.rs]: Removed

File: blockchain/message_pool/src/msgpool/msg_pool.rs (hunk 20 above)
Comment (tyshko5) [reply blockchain/message_pool/src/msgpool/msg_pool.rs]: Removed

File: forest/daemon/src/daemon.rs (hunk 32 above)
Comment (tyshko5) [reply forest/daemon/src/daemon.rs]: Fixed

File: forest/daemon/src/daemon.rs (hunk 28 above)
Comment (tyshko5) [reply forest/daemon/src/daemon.rs:406]: Approximately 2.2 years. I fixed our time to be within the limit

File: forest/daemon/src/daemon.rs (hunk 27 above)
Comment (tyshko5) [reply forest/daemon/src/daemon.rs]: Fixed accordingly

File: forest/daemon/src/daemon.rs (hunk 26 above)
Comment (tyshko5) [reply forest/daemon/src/daemon.rs]: Fixed accordingly

File: forest/daemon/src/daemon.rs (hunk 29 above)
Comment (lemmih) [reply forest/daemon/src/daemon.rs:407]: `bail!` is shorter. Doesn't matter much.

File: blockchain/consensus/deleg_cns/src/proposer.rs [hunk 33]
```diff
@@ -102,18 +102,18 @@ impl Proposer for DelegatedProposer {
...
         DB: Blockstore + Store + Clone + Sync + Send + 'static,
         MP: MessagePoolApi + Send + Sync + 'static,
     {
         services.spawn(async move {
-            if let Err(e) = self.run(state_manager, mpool.as_ref(), &submitter).await {
-                error!("block proposal stopped: {}", e)
-            }
+            self.run(state_manager, mpool.as_ref(), &submitter)
+                .await
+                .map_err(|e| anyhow::anyhow!("block proposal stopped: {}", e))
```
Comment (lemmih) [inline blockchain/consensus/deleg_cns/src/proposer.rs:114]: Can `context` be used here?

File: blockchain/consensus/deleg_cns/src/proposer.rs [hunk 34]
```diff
@@ -102,18 +102,18 @@ impl Proposer for DelegatedProposer {
...
         });
-        Ok(())
+        anyhow::Ok(())
```
Comment (lemmih) [inline blockchain/consensus/deleg_cns/src/proposer.rs:116]: Is this necessary?

File: forest/daemon/src/daemon.rs [hunk 35]
```diff
@@ -143,7 +145,7 @@ pub(super) async fn start(config: Config, detached: bool) -> Db {
 
     let db = open_db(&config);
 
-    let mut services = JoinSet::new();
+    let mut services: JoinSet<anyhow::Result<()>> = JoinSet::new();
```
Comment (lemmih) [inline forest/daemon/src/daemon.rs:148]: Is this type signature necessary?

File: forest/daemon/src/daemon.rs [hunk 36]
```diff
@@ -392,6 +393,22 @@ pub(super) async fn start(config: Config, detached: bool) -> Db {
     db
 }
 
+async fn propagate_error(services: &mut JoinSet<Result<(), anyhow::Error>>) -> anyhow::Error {
```
Comment (lemmih) [inline forest/daemon/src/daemon.rs:396]: Document this function.

File: forest/daemon/src/daemon.rs (hunk 28 above)
Comment (lemmih) [reply forest/daemon/src/daemon.rs:406]: I don't see your fix. Please push your changes before marking issues as resolved. That makes it much easier for reviewers to keep track of things.

File: forest/daemon/src/daemon.rs (hunk 35 above)
Comment (tyshko5) [reply forest/daemon/src/daemon.rs]: Removed

File: blockchain/consensus/deleg_cns/src/proposer.rs (hunk 34 above)
Comment (tyshko5) [reply blockchain/consensus/deleg_cns/src/proposer.rs:116]: Yes, as `Proposer` trait requires `anyhow::Result` as return type

File: blockchain/consensus/deleg_cns/src/proposer.rs (hunk 33 above)
Comment (tyshko5) [reply blockchain/consensus/deleg_cns/src/proposer.rs]: Fixed

File: forest/daemon/src/daemon.rs [hunk 37]
```diff
@@ -377,8 +376,10 @@ pub(super) async fn start(config: Config, detached: bool) -> Db {
 
     services.spawn(p2p_service.run());
 
-    // Block until ctrl-c is hit
-    ctrlc_oneshot.await.unwrap();
+    select! {
```
[the remark this answers is not in the document]
Comment (tyshko5) [reply forest/daemon/src/daemon.rs:381]: Done

File: forest/daemon/src/daemon.rs (hunk 36 above)
Comment (tyshko5) [reply forest/daemon/src/daemon.rs:401]: Done

**Review by lemmih — APPROVED** [review]
It would be nice to return error code 1 when something goes wrong but let's do that in another PR.

**Review by hanabi1224 — APPROVED** [review]

**Comment by lemmih** [comment]
> In which situations are we entering this code:
> 
> ```
>     tokio::time::sleep(Duration::new(64000000, 0)).await;
>     anyhow::Error::msg(
>         "More than 2 years have passed, all services are down and Forest was still running",
>     )
> ```
> 
> I'm not sure to understand the purpose of it.

The function should wait for errors. If there are no running services then it should wait forever. But there's no function for sleeping forever in tokio, I believe. So we just wait 2 years instead. Some code comments to explain this would be nice. Could even loop forever to avoid problems if someone wants to run Forest for more than 2 years.

**Comment by elmattic** [comment]
> If there are no running services then it should wait forever.

That's what I don't understand, in which circumstances can you end up in a state in no services at all have been spawn?
We should always have for instance `ChainMuxer` no?

**Comment by lemmih** [comment]
> > If there are no running services then it should wait forever.
> 
> That's what I don't understand, in which circumstances can you end up in a state in no services at all have been spawn? We should always have for instance `ChainMuxer` no?

Right now, yes, but that may not be true in the future. One could possibly imagine that the light-client wouldn't run any background services.

**Review by lemmih — APPROVED** [review]

File: forest/daemon/src/daemon.rs [hunk 38]
```diff
@@ -392,6 +395,27 @@ pub(super) async fn start(config: Config, detached: bool) -> Db {
...
+                }
+            },
+        }
+    }
+    // In case all services are down without errors we are still willing
+    // to wait indefinitely for CTRL-C signal. As `tokio::time::sleep` has
+    // a limit of approximately 2.2 years we have to loop
+    loop {
+        tokio::time::sleep(Duration::new(64000000, 0)).await;
+    }
```
Comment (LesnyRumcajs) [inline forest/daemon/src/daemon.rs:416]: I think after 2 years of waiting we are fine to retire this process...

File: forest/daemon/src/daemon.rs [hunk 39]
```diff
@@ -392,6 +395,27 @@ pub(super) async fn start(config: Config, detached: bool) -> Db {
     db
 }
 
+// returns the first error with which any of the services end
+// in case all services finished without an error sleeps for more than 2 years
+// and then returns with an error
```
Comment (LesnyRumcajs) [inline forest/daemon/src/daemon.rs:400]: this is not true, no? I mean, if I understand the code correctly, then if all services finished without an error then nothing is returned, ever. Or am I missing something?

File: forest/daemon/src/daemon.rs (hunk 38 above)
Comment (tyshko5) [reply forest/daemon/src/daemon.rs:416]: David explains in the upper comments that a situation in which we should wait indefinitely

File: forest/daemon/src/daemon.rs (hunk 39 above)
Comment (tyshko5) [reply forest/daemon/src/daemon.rs:400]: Yes, we are waiting for CTRL-C to be pressed, otherwise we are waiting indefinitely

**Pull request merged**